pygeos/constructive.py

```diff
@@ -486,6 +487,36 @@ def point_on_surface(geometry, **kwargs):
     return lib.point_on_surface(geometry, **kwargs)
 
 
+def polygonize(geometries, **kwargs):
+    """Creates polygons formed from the linework of a set of Geometries.
+
+    Polygonizes an array of Geometries that contain linework which
+    represents the edges of a planar graph. All types of Geometry are
+    accepted as input; the constituent linework is extracted as the edges
```

Comment:
Suggestion:
```
Any type of Geometry may be provided as input; only lines 
will be used to create the output polygons.
```

Reply:
@brendan-ward the original text is certainly not the best worded, but I think "only lines will be used to create output polygons" is a bit misleading, since all the rings (i.e. lines) from polygons will be used as well

Reply:
Perhaps "only lines or rings..." would clarify?

Mostly what I was going for was to try and clarify that any types can be passed in but only parts of them are used; just a slightly different phrasing to in particular change "all types" to "any type".

Reply:
It would be good to add a bit more to the docstring to clarify output behavior, something like:
```
Lines or rings that when combined do not completely close a polygon 
will result in an empty GeometryCollection.  Duplicate segments are 
ignored.
```
(this happens for the first one or two of your example lines below)

and 

```
Any input arrays of lines or rings must have the same shape: each 
polygon to be created must have the same number input lines or 
rings.
```
If you try to pass in a ragged array (e.g., array with first subarray 1 element long and second subarray 2 elements long), you get back `TypeError: One of the arguments is of incorrect type. Please provide only Geometry objects.` which is not particularly helpful in this case.  Perhaps since that may be a common error, specifically validating shape of the inputs would be helpful?

Reply:
> Perhaps since that may be a common error, specifically validating shape of the inputs would be helpful?

We can't simply check the shape of the input, since you can also pass a 1D array to get a scalar output. So for a 1D array, we would need to check the actual values (whether it are geometries or arrays), which will become a bit costly I think.

Reply:
I already updated the docstring for your first suggestions about only lines/rings being used as input. 

I am still struggling a bit to find a good wording this suggestion:

```
Any input arrays of lines or rings must have the same shape: each 
polygon to be created must have the same number input lines or 
rings.
```

I also find that a bit confusing, since its sounds as it is talking about multiple arrays, while it's actually about a 2D array?

(now, in general I don't think the fact that this is a ufunc is going to be very useful. I expect that most people will only use this with a 1D array)

Reply:
Feel free to ignore any confusing suggestions. 😄 

We can always refine the docstring later as we better identify where users are getting confused.  Don't let the docstring suggestions here hold things up.

Reply:
Rereading your previous comment, I think I now better understand the point you were trying to make. 
Probably easiest to see if you pass a list of arrays/lists, and for example using one of the other constructive functions:

```
In [4]: pygeos.linestrings([[(0, 1), (2, 3)], [(0, 1), (2, 3)]])
Out[4]: 
array([<pygeos.Geometry LINESTRING (0 1, 2 3)>,
       <pygeos.Geometry LINESTRING (0 1, 2 3)>], dtype=object)

In [5]: pygeos.linestrings([[(0, 1), (2, 3)], [(0, 1), (2, 3), (4, 5)]])
...
ValueError: linestrings: Input operand 0 does not have enough dimensions (has 1, gufunc core with signature (i, d)->() requires 2)
```

A different error message, but also not an informative one, and similar in the underlying reason: if you pass a '"ragged" array, it's not a proper multidimensional array but rather a 1D array of wrong objects, and you get an error. 

But since the other constructive functions have a similar "trap", maybe can leave this for a separate issue to think about how we can make this more user friendly? 
(and then will merge this, as I have a follow-up on this for `polygonize_full`)